adding support for user selection of jar or war files for mvm deployment

etanshaul opened the pull request:
fixes #449 

A possible approach to allowing the user to manually select jar or war files for mvm deployment. Gets around forcing the user to have preconfigured IJ Artifacts.

The other approach discussed was to use IJ Artifact system to create a custom Artifact pointing to a file on the filesystem for deployment. 

todo and better documentation pending

Updated UI and placeholder deployment source selected:
<img width="549" alt="screen shot 2016-03-15 at 11 16 03 pm" src="https://cloud.githubusercontent.com/assets/1735744/13801053/d7489b18-eb04-11e5-987a-df4f4eaed8d8.png">

A custom chosen jar file:
<img width="553" alt="screen shot 2016-03-15 at 11 19 56 pm" src="https://cloud.githubusercontent.com/assets/1735744/13801068/f1b2a674-eb04-11e5-89ae-f2c7f1a34364.png">

Deployment source dropdown:
<img width="331" alt="screen shot 2016-03-15 at 11 20 12 pm" src="https://cloud.githubusercontent.com/assets/1735744/13801079/00e60cc6-eb05-11e5-9864-336d1b734ce8.png">

patflynn commented:
Thanks for sending this out. The screenshots answer all my questions. Super useful.

My comments are:
- 'Managed VM Deployment' is a confusing deployment name since whatever you select will be a managed vm deployment. 
- If an artifact is selected, is the 'use custom deployment file' still visible?
- 'use custom deployment file' should be something like 'Use alternate deployment archive', and filter by jars and wars.

elharo commented:
looks good overall. I suggest trying to align the left side of the name input box with the dropdowns below it.

etanshaul commented:
@patflynn Addressing your points below:

**"'Managed VM Deployment' is a confusing deployment name since whatever you select will be a managed vm deployment."**
Yes this is true, I forgot to include one screenshot that shows why I chose this (I originally had something like "Select a deployment source"):

<img width="376" alt="screen shot 2016-03-16 at 11 34 03 am" src="https://cloud.githubusercontent.com/assets/1735744/13818102/0525b960-eb6b-11e5-9db5-4fff9a8490bf.png">

Notice that the name that is used for the dropdown item also shows up under the Application Servers panel if the user keeps this placeholder option selected together with a manually browsed to jar/war file. The reason I made this placeholder in the first place is that our extra settings are only visible if there is a deployment source in the dropdown. Therefore I had to create a "fake" deployment source (I updated the icon to make it look more informational) to act as a placeholder.

**"If an artifact is selected, is the 'use custom deployment file' still visible?"**
Yes. One thing I was going to try and do is disable the text area in this case - I prefer this to hiding and showing the field which messes with vertical flow of the dialog.

**'use custom deployment file' should be something like 'Use alternate deployment archive', and filter by jars and wars.**
Yes, that is better.

etanshaul commented:
One other caveat I forgot to mention:

The elements in the deployment dropdown menu are sorted (by the IJ framework code) alphabetically - its a `SortedComboBoxModel` using a `String.CASE_INSENSITIVE_ORDER` comparator. 

What this means to us is that depending on the label we choose for that placeholder I discussed, it could appear _in the middle_ of the dropdown list. This is not ideal, but it is a consequence of the fact that we are "forcing" our workflow on the IJ ui.

etanshaul commented:
@elharo those UI elements you referenced are part of the IJ framework not exposed via an extension point. The custom part of that UI is the manual deployment file selection as well as the custom yaml docker file selection.

etanshaul commented:
I've pushed a couple updates including wording improvements. Before I finalize all the polish on the PR I wanted to make sure we were on board with the approach? My opinion is that it is a lot better than the other artifact based approach I showed (using an IJ artifact to act as a pointer to a filesystem archive).

patflynn commented:
I think I'm on board. My hesitation is around the way the artifact is displayed in that application servers screenshot you included.  I need to spend a bit of time playing with the capabilties there to make sure we don't break that model.

etanshaul commented:
It just hit me that there is a way around what is displayed in that Application Server window you mentioned. Its not pretty - involved making a field mutable that wasn't intended to be, as well as instanceof and a cast - but the UI looks a lot better with it:

![image](https://cloud.githubusercontent.com/assets/1735744/13850145/211a1f66-ec2f-11e5-8cb0-9254c2494045.png)

When the user manually browses to an archive, I am setting the displayed name to be the file name.

patflynn commented:
sweet.

On Thu, Mar 17, 2016 at 11:01 AM, Etan Shaul <notifications@github.com>
wrote:

> It just hit me that there is a way around what is displayed in that
> Application Server window you mentioned. Its not pretty - involved making a
> field mutable that wasn't intended to be, as well as instanceof and a cast
> - but the UI looks a lot better with it:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/1735744/13850145/211a1f66-ec2f-11e5-8cb0-9254c2494045.png>
>
> When the user manually browses to an archive, I am setting the displayed
> name to be the file name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/505#issuecomment-197922085>
>

etanshaul commented:
Minor update:

Mutating the name of the deployment source like I was doing cause the placeholder dropdown element (see previous screenshot) to immediately show the new name - this is not what we want: it doesn't make sense for it to display the name of the browsed to archive.  As a compromise I settled with a hybrid display:

![image](https://cloud.githubusercontent.com/assets/1735744/13883452/e270b0b4-ecff-11e5-9de6-13b4c17151f5.png)

This way the name still makes sense when viewed from the context of the dropdown.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmDeploymentConfiguration.java:
```diff
@@ -48,8 +50,20 @@ public String toString() {
 
   private String dockerFilePath;
   private String appYamlPath;
+  private boolean manualDeployment;
```
how about userSpecifiedArtifactPath?

etanshaul replied:
I was actually just trying to come up with better naming myself - yea thats much better. Also, I've iterated on the UI again improving some things. Will update screenshots.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmDeploymentRunConfigurationEditor.java:
```diff
@@ -58,11 +61,20 @@
   private JButton generateAppYamlButton;
   private JButton generateDockerfileButton;
-
+  private JCheckBox manualDeploymentFileCheckbox;
```
I prefer userSpecifiedArtifactPath. LMKWYT

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmCloudType.java:
```diff
@@ -148,6 +159,36 @@ public ManagedVmDeploymentConfiguration createDefaultConfiguration(
               server.getConfiguration().getCloudProjectName()));
     }
+
+    protected static class PlaceholderDeploymentSource extends ModuleDeploymentSourceImpl {
```
if we go the UserSpecified.. naming route you could also rename this class to UserSpecfiedPathDeploymentSource or some shit like that.

etanshaul replied:
In my current iteration (still not pushed) I had updated this to FileSystemDeploymentSource. 

But I like your wording better.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmDeploymentRunConfigurationEditor.java:
```diff
@@ -76,6 +88,12 @@ public void actionPerformed(ActionEvent e) {
         }
       }
     });
+    manualDeploymentFileSelector.addBrowseFolderListener(
```
is there a way to filter the list of files in the browser by some predicate?

etanshaul commented:
New screenshots reflecting updated UX:

_Deploy window with a preconfigured IJ Artifact Selected:_
<img width="555" alt="screen shot 2016-03-20 at 5 04 24 pm" src="https://cloud.githubusercontent.com/assets/1735744/13907132/fca83bcc-eebd-11e5-8d0f-6a99089ecc78.png">

_Deploy window with option to browse to user selected archive:_
<img width="560" alt="screen shot 2016-03-20 at 5 04 34 pm" src="https://cloud.githubusercontent.com/assets/1735744/13907133/003c0cc8-eebe-11e5-82a2-5f9ae0355b52.png">

_Closeup of deployment options:_
<img width="340" alt="screen shot 2016-03-20 at 5 04 41 pm" src="https://cloud.githubusercontent.com/assets/1735744/13907134/03a1d3fc-eebe-11e5-8346-b30a4426d014.png">

_Application Servers window after selecting to deploy a user chosen archive:_
<img width="466" alt="screen shot 2016-03-20 at 5 05 06 pm" src="https://cloud.githubusercontent.com/assets/1735744/13907135/07bfa054-eebe-11e5-8646-1cffd9dfff26.png">

patflynn commented:
Amazing.

On Sun, Mar 20, 2016 at 5:08 PM, Etan Shaul <notifications@github.com>
wrote:

> New screenshots reflecting updated UX:
>
> *Deploy window with a preconfigured IJ Artifact Selected:*
> [image: screen shot 2016-03-20 at 5 04 24 pm]
> <https://cloud.githubusercontent.com/assets/1735744/13907132/fca83bcc-eebd-11e5-8d0f-6a99089ecc78.png>
>
> *Deploy window with option to browse to user selected archive:*
> [image: screen shot 2016-03-20 at 5 04 34 pm]
> <https://cloud.githubusercontent.com/assets/1735744/13907133/003c0cc8-eebe-11e5-82a2-5f9ae0355b52.png>
>
> *Closeup of deployment options:*
> [image: screen shot 2016-03-20 at 5 04 41 pm]
> <https://cloud.githubusercontent.com/assets/1735744/13907134/03a1d3fc-eebe-11e5-8346-b30a4426d014.png>
>
> *Application Servers window after selecting to deploy a user chosen
> archive:*
> [image: screen shot 2016-03-20 at 5 05 06 pm]
> <https://cloud.githubusercontent.com/assets/1735744/13907135/07bfa054-eebe-11e5-8646-1cffd9dfff26.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/505#issuecomment-199025597>
>

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmDeploymentRunConfigurationEditor.java:
```diff
@@ -58,11 +64,18 @@
-
+  private JPanel userSpecifiedArtifactPanel;
+  private TextFieldWithBrowseButton userSpecifiedArtifactFileSelector;
+  DeploymentSource deploymentSource;
```
why isn't this private?

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmDeploymentRunConfigurationEditor.java:
```diff
@@ -114,9 +140,56 @@ protected void resetEditorFrom(ManagedVmDeploymentConfiguration configuration) {
+    validateConfiguration();
+  }
+
+  private void toggleJarWarSelector() {
```
toggle to me means flipping the bit, but this doesn't necessarily do that.  Maybe just 'updateJarWarSelector()'?

etanshaul replied:
That works. To add another option: `toggleJarWarSelector(boolean visible)`. "toggle" seems to work a little better this way.

etanshaul replied:
I like your suggestion better actually. Updated the branch with both of your comments.

patflynn commented:
lgtm
